### DISCUSSION TITLE
Auto-adapted calling conventions for avoiding stack overhead

### DISCUSSION THREAD

**Discussion by WhiteBlackGoose** [dotnet/runtime#61249] [Ideas]
Hello. Playing with codegen and structs I noticed that one of the overheads a call introduces is working with the arguments. Consider a case:

```cs
using System.Runtime.CompilerServices;

static int Add(int a, int b) => AddFields(new(){ a = a, b = b });

[MethodImpl(MethodImplOptions.NoInlining)]
static int AddFields(Quack q) => q.a + q.b;

public struct Quack
{
    public int a;
    public int b;
}
```

([sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsAChl1JOVYtyAQQAmugBSaM5bGg1bgASnIBeAHw79AMVYxBu3AaYwA7gasAb1M7U3NgUIiAXytFFSUAbQBZGAwACwhdAEl+QQMU9MycvkEAeT42CCZcGgA5CCymQU1NAHMrAF1VKnVjJ11Xd08DAEUGbDAAa3IARxsHWZpscgBqReA41QBmclwMKAYwEzGJyeVA5XIryh2+7Djrm4sTDeUooA))

In the caller, it first populates ram with the struct's fields, then packs that 8-byte piece of memory into `rcx`.
On the callee side, it packs `rcx` into ram back, and then assigns its 4-byte piece to `eax` (and then reads from ram again to `add`).

Instead, it would be nice to have the caller and the callee to "agree" that the arguments are passed in `eax` and `ecx`, then it would look *much* simplier.

Caller:
```asm
mov ecx, ecx  ; the caller is a function itself, so it takes one of its arguments into the first argument of the callee. But that's a coincedence
mov eax, edx
call AddFields
```

Callee:
```asm
add eax, ecx
ret
```

That's not an existing calling convention. So the idea here is to introduce special callconvs for some functions, to speed them up a bit. I'm not a RyuJIT expert, but I'd assume that it would be like this: we make a special bit of metadata for each function, where we store if the function uses a custom callconv. If it does, there will be a reference to table, like this:

```
// this one is used in the example above
conv_0
arg0 -> eax
arg1 -> ecx

// this one is useful when we need to pass the argument by address, so no reason to pass it in registers
conv_1
arg0 -> push

etc.
```

It will be quite light-weight as there are not too many combinations (because most of combinations can be simplified to just one with a slight argument reorder or other minor adjustments).

Of course, it will be disabled for native export scenarios (and other where our functions are exposed outside of the current jit session).

What do you think?

**Comment by huoyaoyuan** [comment]
I think it can confuse a lot for diagnostic tools.

And also, when modifying calling convention can benefit, it should likely be more beneficial to inline the callee.
